tools/README.md

```diff
@@ -0,0 +1,391 @@
+
+This will pin all dependencies *of these libraries* too.
+
+Run the following to update the `requirements.txt`:
+
+```console
+
+$ pip install pip-tools
+$ pip-compile --generate-hashes tools/sometools/requirements.txt
+
+```
+
+### Add the Python tool
+
+For the purpose of this example we will add a trivial tool that dumps information
+about a Python package as `yaml`
+
+Create a file `tools/sometools/mytool.py` with the following content:
+
+
+```python
+#!/usr/bin/env python3
+
+import sys
+
+import requests
+import yaml
+
+
+def main(*args) -> int:
+    sys.stdout.write(
+        yaml.dump(
+            requests.get(f"https://pypi.python.org/pypi/{args[0]}/json").json()))
+    return 0
+
+
+if __name__ == "__main__":
+    sys.exit(main(*sys.argv[1:]))
+```
+
+### Create the `BUILD` file for `sometools`
+
+If you are adding a tool to an existing toolset you may be able skip this step -
+just make sure that `load` lines are present.
+
+Add the following content to the file `tools/sometools/BUILD`
+
+```starlark
+load("//tools/base:envoy_python.bzl", "envoy_py_binary")
+load("@sometools_pip3//:requirements.bzl", "requirement")
+
+licenses(["notice"])  # Apache 2
+```
+
+Note the loading of `requirement` from `@sometools_pip3`, and the loading of `envoy_py_binary`.
+
+We will use these in the next section.
+
+### Create the bazel target for the tool
+
+Add `mytool.py` as an `envoy_py_binary` to the `tools/sometools/BUILD` file.
+
+This will make the `mytool.py` file runnable as a `bazel` target, and will
+add a test runner to ensure the file is tested.
+
+```starlark
+envoy_py_binary(
+    name = "tools.sometools.mytool",
+    srcs = ["mytool.py"],
+    visibility = ["//visibility:public"],
+    deps = [
+        requirement("requests"),
+        requirement("pyyaml"),
+    ],
+)
+```
+
+Note that the `envoy_py_binary` expects the full dotted name of the module - in this case,
+`tools.sometool.mytool`.
+
+This will create a runnable target with the name of just `mytool`.
+
+With this added users that have `bazel` installed can run the tool with the following command:
+
+```console
+$ bazel run //tools/sometools:mytool PACKAGENAME
+```
+
+
+### Make the tool runnable without bazel
+
+If you want users to be able to run the tool directly without `bazel`, you will need
+to make it executable:
+
+```console
+$ chmod +x tools/sometools/mytool.py
+```
+
+With this, users that have the necessary Python dependencies locally installed
+can run the tool directly with the following command:
+
+```console
+$ ./tools/sometools/mytool.py PACKAGENAME
+
+```
+
+### Add unit tests for the tool
+
+Envoy tooling is tested with `pytest`.
+
+The test runner expects a test file, in this case `tools/sometools/tests/test_mytool.py`.
+
+First, create the required directory if it is not present.
+
+```console
+$ mkdir tools/sometools/tests
+```
+
+Now add the following content to the `tools/sometools/tests/test_mytool.py` file:
```

Comment:
Do we need to describe how the BUILD rule looks?

Reply:
comment above on this